### PR TITLE
Add stream IDs, proper storage prefixing, recovering from crashes.

### PR DESCRIPTION
**Pull request by cube2222**

### CODE REVIEW

File: execution/engine.go [hunk 1]
```diff
@@ -133,37 +133,37 @@ func (engine *PullEngine) Run(ctx context.Context) {
 }
 
 func (engine *PullEngine) loop(ctx context.Context, tx storage.StateTransaction) error {
-	sourcePrefixedTx := tx.WithPrefix(engine.sourceStoragePrefix)
+	prefixedTx := tx.WithPrefix(engine.nodeStatePrefix)
 
-	watermark, err := engine.watermarkSource.GetWatermark(ctx, sourcePrefixedTx)
+	watermark, err := engine.watermarkSource.GetWatermark(ctx, tx)
```
Comment (JasiekChomiak) [inline execution/engine.go]: This seems like a very subtle change, yet there is nothing that explains it. Could you add a comment, as to explain why the watermark is suddenly created using the original transaction, and not the prefixed one, as was in the previous version?

File: execution/engine.go [hunk 2]
```diff
@@ -133,37 +133,37 @@ func (engine *PullEngine) Run(ctx context.Context) {
...
 	if err != nil {
 		return errors.Wrap(err, "couldn't get current watermark from source")
 	}
 	if watermark.After(engine.lastCommittedWatermark) {
-		err := engine.irs.UpdateWatermark(ctx, tx, watermark)
+		err := engine.irs.UpdateWatermark(ctx, prefixedTx, watermark)
```
Comment (JasiekChomiak) [inline execution/engine.go]: Same thing here - these changes need commenting in my opinion

File: execution/engine.go (hunk 1 above)
Comment (cube2222) [reply execution/engine.go]: I've added necessary comments.

File: execution/engine.go (hunk 2 above)
Comment (cube2222) [reply execution/engine.go]: I've added necessary comments.

**Comment by Qoziyev** [comment]
#329

**Pull request merged**